release-21.2: rowexec: fix zigzag joiner with ON expression in some cases

yuzefovich opened the pull request:
Backport 1/1 commits from #71226.
Backport 1/1 commits from #71289.

/cc @cockroachdb/release

---

Zigzag joiner needs to tell the row fetcher which columns are needed.
Previously, we forgot to include the columns that are needed by the ON
expression but are not needed in the output, so when evaluating such an
ON expression, we would hit an internal error. This commit fixes the
problem by including all columns referenced by the ON expression into
the set of columns to be fetched.

Fixes: #71093

Release note (bug fix): Previously, CockroachDB could encounter an
internal error when executing a zigzag join in some cases (when there
are multiple filters present and at least one filter refers to the
column that is part of STORING clause of the secondary index that is
used by the zigzag join), and this has been fixed.

yuzefovich committed:
rowexec: fix zigzag joiner with ON expression in some cases

Zigzag joiner needs to tell the row fetcher which columns are needed.
Previously, we forgot to include the columns that are needed by the ON
expression but are not needed in the output, so when evaluating such an
ON expression, we would hit an internal error. This commit fixes the
problem by including all columns referenced by the ON expression into
the set of columns to be fetched.

Release note (bug fix): Previously, CockroachDB could encounter an
internal error when executing a zigzag join in some cases (when there
are multiple filters present and at least one filter refers to the
column that is part of STORING clause of the secondary index that is
used by the zigzag join), and this has been fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71253)
<!-- Reviewable:end -->

yuzefovich commented:
Will wait for 21.2.1.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71253#-:-MlPd9uaCQ0AbAMklle8:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach)



<!-- Sent from Reviewable.io -->

yuzefovich committed:
rowexec: fix handling of needed cols by zigzag joiner

We recently merged a fix so that the ON expr is handled correctly by the
zigzag joiner. However, that commit had a problem: namely, we forgot to
check that a particular IndexedVar is used, so we could include columns
that aren't actually needed (or even in the index) to be decoded. This
is now fixed.

Additionally, I originally thought that we had a similar problem in the
join reader, but everything turned out to be correct there. This commit
refactors the code there a bit to make things more clear (previously,
we were relying on the IndexedVar's ordinal to be 0 when the indexed var
is unused, so that when we shift down the ordinal by a positive number,
we get a negative ordinal, and we wouldn't include the corresponding
column into the "needed cols" set).

Release note: None (no release with this bug)

cucaroach reviewed (approved):
Presumably this will be re-cast to include the late breaking tweak?   <img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  otherwise

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71253#-:-MlV6k4J5jUJShfnWVtx:bhuvwfr)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71253#-:-Mlm44qTB_L61AGCauci:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->
